### DISCUSSION TITLE
[Feature] Semantic Chunking

### DISCUSSION THREAD

**Discussion by do-me** [do-me/SemanticFinder#45] [Ideas]
I saw [this post](https://www.linkedin.com/posts/pramodith_rag-generativeai-activity-7169676809263915008-KDfz/) about semantic chunking using sliding chunk windows and think this is a cool approach.

In a nutshell you use sliding chunk windows and calculate their embeddings and their distance to each other. Using this score you are trying to identify semantic "break points" i.e. where you have the largest delta. 

Together with minimum and maximum chunk size that would make for a great alternative to the current chunking functions!

Langchain implemented this already and gives some [good explanation in their docs](https://python.langchain.com/docs/modules/data_connection/document_transformers/semantic-chunker).

**Comment by do-me** [comment]
[RAPTOR](https://www.linkedin.com/posts/pramodith_rag-generatieveai-activity-7171453279669858306-egVL?utm_source=share&utm_medium=member_desktop) might also be a good choice. The idea is to create a tree-like structure from chunks by using dimensionality reduction & e.g. k-means for clustering. Afterwards these clusters are clustered again and so on.

Copying from LinkedIn: 

> 
> At retrieval time, the authors propose two algorithms:
> 
> 1. Tree Traversal: Identify the top-k chunks at each level of the tree and send them to the LLM. So if we have a tree of depth 2 and we set k = 3, we’d send 6 chunks to the LLM, 3 from each level.
> 
> 2. Collapsed Tree Traversal → Kind of like regular RAG, the tree is flattened and we retrieve chunks until a specific token limit is reached to maximize the use of the LLM’s context window.
> 
> For more read: https://lnkd.in/eXxwX8xK

It might be a good choice too to combine both ideas, e.g. use semantic chunking (which seems better than relying on error-prone dim reduction) and create a tree-like structure from these chunks instead.

**Comment by do-me** [comment]
@VarunNSrivastava do you maybe have other ideas too or did you see any promising demos or similar?

**Comment by varunneal** [comment]
Interesting! I'll dig into these demos. I spent a lot of time with a very similar approach, and found it a bit uneffective and computationally expensive. I found that breaking at punctuation was a bit more effective...

**Reply by do-me** [reply]
Awesome, great that you have some experience in that already. Super curious about your insights on this!

**Comment by do-me** [comment]
Linking current plans: 
- https://github.com/benbrandt/text-splitter/issues/152
- https://github.com/users/benbrandt/projects/2/views/1?pane=issue&itemId=60813549
